File: kweblens-ui/src/columnParity.test.ts
import { existsSync, readFileSync, writeFileSync } from 'node:fs';
import { dirname, resolve } from 'node:path';

import { afterAll, beforeAll, describe, expect, it, vi } from 'vitest';

import { columnsFor, printerColumnDefs } from './columns';
import type { KubeObject, PrinterColumn } from './types';

/**
 * The SPA half of the column-parity gate.
 *
 * This file is the *producer* of `column-parity/expected.json` and the first of its two
 * consumers; `kweblens-core`'s `ColumnParityTest` is the other. Neither side hand-writes an
 * expectation — the golden is rendered by the code below, which is the only reason comparing it
 * with a second implementation proves anything at all. See `column-parity/README.md`.
 */

/**
 * The corpus lives at the repository root so both modules reach it by the same name. It is found
 * by walking up rather than by a fixed `../..`, because vitest's working directory is not the
 * same when it is run by npm, by maven and by an editor — and a path that resolves to nothing
 * would make this whole file silently skip.
 */
function corpusDir(): string {
  let dir = process.cwd();
  for (let up = 0; up < 6; up++) {
    if (existsSync(resolve(dir, 'column-parity/objects.json'))) {
      return resolve(dir, 'column-parity');
    }
    dir = dirname(dir);
  }
  throw new Error(`column-parity/ not found above ${process.cwd()}`);
}

const CORPUS = corpusDir();

interface Corpus {
  now: string;
  statusColumns: Record<string, string[]>;
  cases: { kind: string; name: string; object: KubeObject }[];
  printerColumns: { name: string; columns: PrinterColumn[]; object: KubeObject }[];
}

interface Golden {
  cases: { kind: string; name: string; values: Record<string, string> }[];
  printerColumns: { name: string; values: Record<string, string> }[];
}

const corpus = JSON.parse(readFileSync(resolve(CORPUS, 'objects.json'), 'utf8')) as Corpus;

/** The keys a kind's values are recorded under: every column but the server's own state. */
function keysFor(kind: string): string[] {
  const excluded = new Set(corpus.statusColumns[kind] ?? []);
  return columnsFor(kind)
    .map((c) => c.key)
    .filter((k) => !excluded.has(k));
}

function render(): Golden {
  return {
    cases: corpus.cases.map((c) => {
      const columns = columnsFor(c.kind);
      const wanted = new Set(keysFor(c.kind));
      const values: Record<string, string> = {};
      for (const column of columns) {
        if (wanted.has(column.key)) {
          values[column.key] = column.render(c.object);
        }
      }
      return { kind: c.kind, name: c.name, values };
    }),
    printerColumns: corpus.printerColumns.map((c) => {
      const values: Record<string, string> = {};
      for (const column of printerColumnDefs(c.columns)) {
        values[column.key] = column.render(c.object);
      }
      return { name: c.name, values };
    }),
  };
}

describe('column parity corpus', () => {
  // Every `date`-typed printer column measures against the wall clock, so the golden is only a
  // fact if the clock is one. `now` travels in the corpus so the Java side can be handed the
  // same instant rather than a second guess at it.
  beforeAll(() => {
    vi.useFakeTimers();
    vi.setSystemTime(new Date(corpus.now));
    // Regeneration happens before either assertion, so one run with UPDATE_COLUMN_PARITY set
    // leaves a golden both tests then agree with — rather than one that the coverage check,
    // which reads the file, was still comparing against in its previous state.
    if (process.env.UPDATE_COLUMN_PARITY) {
      writeFileSync(resolve(CORPUS, 'expected.json'), JSON.stringify(render(), null, 2) + '\n');
    }
  });
  afterAll(() => {
    vi.useRealTimers();
  });

  // Deliberately reads the golden FILE rather than what render() just produced: comparing
  // render() with itself would pass for any column set at all. What this asserts is that the
  // golden on disk still describes every column `columns.ts` defines — i.e. that a column added
  // to the SPA reached the corpus, and did not just quietly stop being compared with anything.
  it('the golden covers every column of every kind the corpus names', () => {
    const golden = JSON.parse(readFileSync(resolve(CORPUS, 'expected.json'), 'utf8')) as Golden;
    const kinds = new Set(corpus.cases.map((c) => c.kind));
    expect(kinds.size).toBeGreaterThan(0);
    for (const kind of kinds) {
      const keys = keysFor(kind);
      expect(keys.length, `${kind} defines no columns`).toBeGreaterThan(0);
      const covered = golden.cases.filter((c) => c.kind === kind);
      expect(covered.length, `${kind} has no case in the golden`).toBeGreaterThan(0);
      for (const one of covered) {
        expect(Object.keys(one.values).sort(), `${kind} / ${one.name}`).toEqual([...keys].sort());
      }
    }
  });

  it('renders what the golden records', () => {
    const golden = JSON.parse(readFileSync(resolve(CORPUS, 'expected.json'), 'utf8')) as Golden;
    expect(render()).toEqual(golden);
  });
});

File: kweblens-core/src/test/java/org/alexmond/kweblens/column/ColumnParityTest.java
package org.alexmond.kweblens.column;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.time.Clock;
import java.time.Instant;
import java.time.ZoneOffset;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.fabric8.kubernetes.api.model.GenericKubernetesResource;
import io.fabric8.kubernetes.client.utils.Serialization;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.TestInstance;

import org.alexmond.kweblens.resource.PrinterColumn;

import static org.assertj.core.api.Assertions.assertThat;

/**
 * The Java half of the column-parity gate: the same objects, through this package, must
 * produce the same strings the SPA produces.
 *
 * <p>
 * <b>Nothing in this file writes an expectation.</b> That is the whole design.
 * {@code column-parity/expected.json} is rendered by {@code columns.ts} in
 * {@code kweblens-ui/src/columnParity.test.ts}, and both sides then assert against it —
 * so a migration that quietly re-decides what {@code Ready} means goes red here, and a
 * change to the SPA goes red there first and here second. Two hand-written expectations
 * that agree would prove only that one author held one idea twice.
 *
 * <p>
 * The other half of the gate is the coverage assertion: the keys this package renders
 * must be exactly the keys the golden carries, so a column added to one side and
 * forgotten on the other is a failure rather than a value nobody compares.
 */
@TestInstance(TestInstance.Lifecycle.PER_CLASS)
class ColumnParityTest {

	private final JsonNode corpus = read("objects.json");

	private final JsonNode golden = read("expected.json");

	@Test
	void everyKindRendersWhatTheSpaRenders() {
		List<String> compared = new ArrayList<>();
		for (int i = 0; i < this.corpus.get("cases").size(); i++) {
			JsonNode declared = this.corpus.get("cases").get(i);
			JsonNode expected = this.golden.get("cases").get(i);
			String kind = declared.get("kind").asText();
			String name = kind + " / " + declared.get("name").asText();
			assertThat(expected.get("kind").asText()).as("corpus and golden out of step at " + name).isEqualTo(kind);
			List<Column> columns = ColumnCatalog.forResourceId(kind);
			assertThat(columns).as("no server-side columns for " + kind).isNotEmpty();
			Map<String, String> rendered = render(columns, object(declared));
			// Keys first, and separately: a column the SPA has and this package does not
			// is
			// the failure most likely to be read as "some value is wrong" from a diff of
			// fifteen strings.
			assertThat(rendered.keySet()).as("columns covered for " + name)
				.containsExactlyInAnyOrderElementsOf(strings(expected.get("values")).keySet());
			assertThat(rendered).as(name).isEqualTo(strings(expected.get("values")));
			compared.add(name);
		}
		assertThat(compared).hasSize(this.golden.get("cases").size());
	}

	@Test
	void everyCoveredKindAppearsInTheCorpus() {
		List<String> inCorpus = new ArrayList<>();
		this.corpus.get("cases").forEach((node) -> inCorpus.add(node.get("kind").asText()));
		assertThat(inCorpus).as("a covered kind with no corpus case is a kind nothing compares")
			.containsAll(ColumnCatalog.coveredResourceIds());
	}

	@Test
	void crdPrinterColumnsRenderWhatTheSpaRenders() {
		Clock clock = Clock.fixed(Instant.parse(this.corpus.get("now").asText()), ZoneOffset.UTC);
		for (int i = 0; i < this.corpus.get("printerColumns").size(); i++) {
			JsonNode declared = this.corpus.get("printerColumns").get(i);
			JsonNode expected = this.golden.get("printerColumns").get(i);
			List<PrinterColumn> columns = new ArrayList<>();
			declared.get("columns")
				.forEach((column) -> columns.add(new PrinterColumn(column.get("name").asText(),
						column.get("jsonPath").asText(), column.get("type").asText())));
			assertThat(render(PrinterColumns.of(columns, clock), object(declared))).as(declared.get("name").asText())
				.isEqualTo(strings(expected.get("values")));
		}
	}

	private static Map<String, String> render(List<Column> columns, GenericKubernetesResource object) {
		Map<String, String> values = new LinkedHashMap<>();
		for (Column column : columns) {
			values.put(column.key(), column.render(object));
		}
		return values;
	}

	private static Map<String, String> strings(JsonNode node) {
		Map<String, String> values = new LinkedHashMap<>();
		node.properties().forEach((entry) -> values.put(entry.getKey(), entry.getValue().asText()));
		return values;
	}

	/**
	 * Through the client's own deserializer, not a bare Jackson mapper — the numeric and
	 * map types a column renderer meets at runtime are the ones fabric8 produced, and a
	 * test that built them another way would be measuring its own fixture loader.
	 */
	private static GenericKubernetesResource object(JsonNode node) {
		return Serialization.unmarshal(node.get("object").toString(), GenericKubernetesResource.class);
	}

	private static JsonNode read(String name) {
		try {
			return new ObjectMapper().readTree(Files.readString(corpus().resolve(name)));
		}
		catch (IOException ex) {
			throw new IllegalStateException("cannot read the parity corpus: " + name, ex);
		}
	}

	/**
	 * The corpus lives at the repository root because it belongs to neither module. Found
	 * by walking up rather than by a fixed {@code ../}, so this test does not depend on
	 * which directory the build happened to start it in — a path that resolved to nothing
	 * would make the whole file pass by never comparing anything.
	 */
	private static Path corpus() {
		Path dir = Paths.get("").toAbsolutePath();
		for (int up = 0; up < 6 && dir != null; up++) {
			Path candidate = dir.resolve("column-parity");
			if (Files.isRegularFile(candidate.resolve("objects.json"))) {
				return candidate;
			}
			dir = dir.getParent();
		}
		throw new IllegalStateException("column-parity/ not found above " + Paths.get("").toAbsolutePath());
	}

}
